File: day4/puzzle1.py
from re import compile


def main():
    pattern = compile(r"\d{1,2}")
    sum = 0
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            _, numbers = line[:-1].split(":")
            winning_numbers, my_numbers = numbers.split("|")
            winning_numbers = [int(n) for n in pattern.findall(winning_numbers)]
            my_numbers = [int(n) for n in pattern.findall(my_numbers)]
            points = 0
            for n in my_numbers:
                if n not in winning_numbers:
                    continue
                if points == 0:
                    points = 1
                else:
                    points *= 2
            sum += points
    print(f"sum = {sum}")


if __name__ == "__main__":
    main()

File: day4/puzzle1.lua
local function main()
    local f = io.open("puzzle_input.txt", "r")
    if not f then
        os.exit(1)
    end
    local sum = 0
    for line in f:lines() do
        local colon = line:find(":")
        local bar = line:find("|")
        local winning_numbers = {}
        for value in line:sub(colon + 1, bar):gmatch("%d+") do
            winning_numbers[tonumber(value)] = true
        end
        local my_numbers = {}
        for value in line:sub(bar + 1, -1):gmatch("%d+") do
            table.insert(my_numbers, tonumber(value))
        end
        local points = 0
        for _, n in ipairs(my_numbers) do
            if winning_numbers[n] then
                if points == 0 then
                    points = 1
                else
                    points = points * 2
                end
            end
        end
        sum = sum + points
    end
    f:close()
    print(string.format("sum = %d", sum))
end

main()
